File: pocketbase/pb_migrations/1752710000_create_collections.go
package migrations

import (
	"encoding/json"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/daos"
	m "github.com/pocketbase/pocketbase/migrations"
	"github.com/pocketbase/pocketbase/models"
)

func init() {
	m.Register(func(db dbx.Builder) error {
		dao := daos.New(db)

		// Create Collections collection
		collectionsCollection := &models.Collection{}
		json.Unmarshal([]byte(`{
			"id": "collections_id",
			"name": "collections",
			"type": "base",
			"system": false,
			"schema": [
				{
					"id": "user_id_field",
					"name": "user_id",
					"type": "text",
					"system": false,
					"required": true,
					"options": {
						"min": 1,
						"max": 255
					}
				},
				{
					"id": "card_id_field",
					"name": "card_id",
					"type": "text",
					"system": false,
					"required": true,
					"options": {
						"min": 1,
						"max": 255
					}
				},
				{
					"id": "quantity_field",
					"name": "quantity",
					"type": "number",
					"system": false,
					"required": true,
					"options": {
						"min": 1
					}
				},
				{
					"id": "condition_field",
					"name": "condition",
					"type": "select",
					"system": false,
					"required": true,
					"options": {
						"values": ["NM", "LP", "MP", "HP", "DMG"]
					}
				},
				{
					"id": "foil_field",
					"name": "foil",
					"type": "bool",
					"system": false,
					"required": false
				},
				{
					"id": "acquired_date_field",
					"name": "acquired_date",
					"type": "date",
					"system": false,
					"required": false
				},
				{
					"id": "notes_field",
					"name": "notes",
					"type": "text",
					"system": false,
					"required": false,
					"options": {
						"max": 1000
					}
				}
			],
			"indexes": [
				"CREATE INDEX idx_collections_user_id ON collections (user_id)",
				"CREATE INDEX idx_collections_card_id ON collections (card_id)",
				"CREATE INDEX idx_collections_user_card ON collections (user_id, card_id)"
			]
		}`), collectionsCollection)

		if err := dao.SaveCollection(collectionsCollection); err != nil {
			return err
		}

		// Create Cards collection
		cardsCollection := &models.Collection{}
		json.Unmarshal([]byte(`{
			"id": "cards_id",
			"name": "cards",
			"type": "base",
			"system": false,
			"schema": [
				{
					"id": "scryfall_id_field",
					"name": "scryfall_id",
					"type": "text",
					"system": false,
					"required": true,
					"options": {
						"min": 1,
						"max": 255
					}
				},
				{
					"id": "name_field",
					"name": "name",
					"type": "text",
					"system": false,
					"required": true,
					"options": {
						"min": 1,
						"max": 255
					}
				},
				{
					"id": "set_code_field",
					"name": "set_code",
					"type": "text",
					"system": false,
					"required": true,
					"options": {
						"min": 1,
						"max": 10
					}
				},
				{
					"id": "set_name_field",
					"name": "set_name",
					"type": "text",
					"system": false,
					"required": true,
					"options": {
						"min": 1,
						"max": 255
					}
				},
				{
					"id": "rarity_field",
					"name": "rarity",
					"type": "select",
					"system": false,
					"required": true,
					"options": {
						"values": ["common", "uncommon", "rare", "mythic", "special", "bonus"]
					}
				},
				{
					"id": "mana_cost_field",
					"name": "mana_cost",
					"type": "text",
					"system": false,
					"required": false,
					"options": {
						"max": 50
					}
				},
				{
					"id": "type_line_field",
					"name": "type_line",
					"type": "text",
					"system": false,
					"required": true,
					"options": {
						"min": 1,
						"max": 255
					}
				},
				{
					"id": "colors_field",
					"name": "colors",
					"type": "json",
					"system": false,
					"required": false
				},
				{
					"id": "image_file_field",
					"name": "image_file",
					"type": "file",
					"system": false,
					"required": false,
					"options": {
						"maxSelect": 1,
						"maxSize": 5242880,
						"mimeTypes": ["image/jpeg", "image/png", "image/webp"]
					}
				},
				{
					"id": "price_usd_field",
					"name": "price_usd",
					"type": "number",
					"system": false,
					"required": false,
					"options": {
						"min": 0
					}
				},
				{
					"id": "last_updated_field",
					"name": "last_updated",
					"type": "date",
					"system": false,
					"required": false
				}
			],
			"indexes": [
				"CREATE UNIQUE INDEX idx_cards_scryfall_id ON cards (scryfall_id)",
				"CREATE INDEX idx_cards_name ON cards (name)",
				"CREATE INDEX idx_cards_set_code ON cards (set_code)",
				"CREATE INDEX idx_cards_type_line ON cards (type_line)",
				"CREATE INDEX idx_cards_name_search ON cards (name COLLATE NOCASE)",
				"CREATE INDEX idx_cards_set_name ON cards (set_name)"
			]
		}`), cardsCollection)

		if err := dao.SaveCollection(cardsCollection); err != nil {
			return err
		}

		// Create Sync_Status collection
		syncStatusCollection := &models.Collection{}
		json.Unmarshal([]byte(`{
			"id": "sync_status_id",
			"name": "sync_status",
			"type": "base",
			"system": false,
			"schema": [
				{
					"id": "data_type_field",
					"name": "data_type",
					"type": "select",
					"system": false,
					"required": true,
					"options": {
						"values": ["cards", "sets", "prices"]
					}
				},
				{
					"id": "last_sync_field",
					"name": "last_sync",
					"type": "date",
					"system": false,
					"required": false
				},
				{
					"id": "status_field",
					"name": "status",
					"type": "select",
					"system": false,
					"required": true,
					"options": {
						"values": ["success", "failed", "in_progress"]
					}
				},
				{
					"id": "records_processed_field",
					"name": "records_processed",
					"type": "number",
					"system": false,
					"required": false,
					"options": {
						"min": 0
					}
				},
				{
					"id": "error_message_field",
					"name": "error_message",
					"type": "text",
					"system": false,
					"required": false,
					"options": {
						"max": 1000
					}
				}
			],
			"indexes": [
				"CREATE UNIQUE INDEX idx_sync_status_data_type ON sync_status (data_type)",
				"CREATE INDEX idx_sync_status_last_sync ON sync_status (last_sync)"
			]
		}`), syncStatusCollection)

		if err := dao.SaveCollection(syncStatusCollection); err != nil {
			return err
		}

		return nil
	}, func(db dbx.Builder) error {
		dao := daos.New(db)

		// Delete collections in reverse order
		collection, _ := dao.FindCollectionByNameOrId("sync_status")
		if collection != nil {
			if err := dao.DeleteCollection(collection); err != nil {
				return err
			}
		}

		collection, _ = dao.FindCollectionByNameOrId("cards")
		if collection != nil {
			if err := dao.DeleteCollection(collection); err != nil {
				return err
			}
		}

		collection, _ = dao.FindCollectionByNameOrId("collections")
		if collection != nil {
			if err := dao.DeleteCollection(collection); err != nil {
				return err
			}
		}

		return nil
	})
}

File: pocketbase/pb_migrations/1752710000_create_collections.js
/// <reference path="../pb_data/types.d.ts" />

migrate((db) => {
    const cardsCollection = new Collection({
        "name": "cards",
        "type": "base",
        "system": false,
        "fields": [
            {
                "name": "scryfall_id",
                "type": "text",
                "required": true,
                "min": 36,
                "max": 36
            },
            {
                "name": "oracle_id",
                "type": "text",
                "required": false,
                "min": 36,
                "max": 36
            },
            {
                "name": "name",
                "type": "text",
                "required": true,
                "min": 1,
                "max": 255
            },
            {
                "name": "lang",
                "type": "text",
                "required": false,
                "min": 2,
                "max": 5
            },
            {
                "name": "released_at",
                "type": "date",
                "required": false
            },
            {
                "name": "layout",
                "type": "select",
                "required": false,
                "values": [
                    "normal", "split", "flip", "transform", "modal_dfc", "meld",
                    "leveler", "saga", "adventure", "planar", "scheme", "vanguard",
                    "token", "double_faced_token", "emblem", "augment", "host",
                    "art_series", "double_sided", "case", "mutate", "reversible_card",
                    "prototype", "class"
                ]
            },
            {
                "name": "highres_image",
                "type": "bool",
                "required": false
            },
            {
                "name": "image_status",
                "type": "select",
                "required": false,
                "values": ["missing", "placeholder", "lowres", "highres_scan"]
            },
            {
                "name": "cmc",
                "type": "number",
                "required": false,
                "min": 0
            },
            {
                "name": "type_line",
                "type": "text",
                "required": true,
                "min": 1,
                "max": 255
            },
            {
                "name": "color_identity",
                "type": "json",
                "required": false
            },
            {
                "name": "colors",
                "type": "json",
                "required": false
            },
            {
                "name": "keywords",
                "type": "json",
                "required": false
            },
            {
                "name": "mana_cost",
                "type": "text",
                "required": false,
                "max": 100
            },
            {
                "name": "oracle_text",
                "type": "text",
                "required": false,
                "max": 2000
            },
            {
                "name": "power",
                "type": "text",
                "required": false,
                "max": 10
            },
            {
                "name": "toughness",
                "type": "text",
                "required": false,
                "max": 10
            },
            {
                "name": "loyalty",
                "type": "text",
                "required": false,
                "max": 10
            },
            {
                "name": "defense",
                "type": "text",
                "required": false,
                "max": 10
            },
            {
                "name": "card_faces",
                "type": "json",
                "required": false
            },
            {
                "name": "legalities",
                "type": "json",
                "required": false
            },
            {
                "name": "games",
                "type": "json",
                "required": false
            },
            {
                "name": "reserved",
                "type": "bool",
                "required": false
            },
            {
                "name": "foil",
                "type": "bool",
                "required": false
            },
            {
                "name": "nonfoil",
                "type": "bool",
                "required": false
            },
            {
                "name": "finishes",
                "type": "json",
                "required": false
            },
            {
                "name": "oversized",
                "type": "bool",
                "required": false
            },
            {
                "name": "promo",
                "type": "bool",
                "required": false
            },
            {
                "name": "reprint",
                "type": "bool",
                "required": false
            },
            {
                "name": "variation",
                "type": "bool",
                "required": false
            },
            {
                "name": "set_id",
                "type": "text",
                "required": false,
                "min": 36,
                "max": 36
            },
            {
                "name": "set_code",
                "type": "text",
                "required": true,
                "min": 1,
                "max": 10
            },
            {
                "name": "set_name",
                "type": "text",
                "required": true,
                "min": 1,
                "max": 255
            },
            {
                "name": "set_type",
                "type": "text",
                "required": false,
                "max": 50
            },
            {
                "name": "collector_number",
                "type": "text",
                "required": false,
                "max": 20
            },
            {
                "name": "digital",
                "type": "bool",
                "required": false
            },
            {
                "name": "rarity",
                "type": "select",
                "required": true,
                "values": [
                    "common",
                    "uncommon",
                    "rare",
                    "mythic",
                    "special",
                    "bonus"
                ]
            },
            {
                "name": "artist",
                "type": "text",
                "required": false,
                "max": 255
            },
            {
                "name": "artist_ids",
                "type": "json",
                "required": false
            },
            {
                "name": "border_color",
                "type": "select",
                "required": false,
                "values": ["black", "white", "borderless", "silver", "gold", "yellow"]
            },
            {
                "name": "frame",
                "type": "text",
                "required": false,
                "max": 20
            },
            {
                "name": "security_stamp",
                "type": "select",
                "required": false,
                "values": ["oval", "triangle", "acorn", "arena", "heart", "circle"]
            },
            {
                "name": "full_art",
                "type": "bool",
                "required": false
            },
            {
                "name": "textless",
                "type": "bool",
                "required": false
            },
            {
                "name": "booster",
                "type": "bool",
                "required": false
            },
            {
                "name": "story_spotlight",
                "type": "bool",
                "required": false
            },
            {
                "name": "edhrec_rank",
                "type": "number",
                "required": false,
                "min": 0
            },
            {
                "name": "penny_rank",
                "type": "number",
                "required": false,
                "min": 0
            },
            {
                "name": "prices",
                "type": "json",
                "required": false
            },
            {
                "name": "price_usd",
                "type": "number",
                "required": false,
                "min": 0
            },
            {
                "name": "price_usd_foil",
                "type": "number",
                "required": false,
                "min": 0
            },
            {
                "name": "price_eur",
                "type": "number",
                "required": false,
                "min": 0
            },
            {
                "name": "price_tix",
                "type": "number",
                "required": false,
                "min": 0
            },
            {
                "name": "related_uris",
                "type": "json",
                "required": false
            },
            {
                "name": "purchase_uris",
                "type": "json",
                "required": false
            },
            {
                "name": "image_uris",
                "type": "json",
                "required": false
            },
            {
                "name": "image_file",
                "type": "file",
                "required": false,
                "maxSelect": 1,
                "maxSize": 5242880,
                "mimeTypes": ["image/jpeg", "image/png", "image/webp"]
            },
            {
                "name": "multiverse_ids",
                "type": "json",
                "required": false
            },
            {
                "name": "mtgo_id",
                "type": "number",
                "required": false,
                "min": 0
            },
            {
                "name": "arena_id",
                "type": "number",
                "required": false,
                "min": 0
            },
            {
                "name": "tcgplayer_id",
                "type": "number",
                "required": false,
                "min": 0
            },
            {
                "name": "cardmarket_id",
                "type": "number",
                "required": false,
                "min": 0
            },
            {
                "name": "preview_info",
                "type": "json",
                "required": false
            },
            {
                "name": "all_parts",
                "type": "json",
                "required": false
            },
            {
                "name": "last_updated",
                "type": "date",
                "required": false
            }
        ],
        "indexes": [
            "CREATE UNIQUE INDEX idx_cards_scryfall_id ON cards (scryfall_id)",
            "CREATE INDEX idx_cards_name ON cards (name)",
            "CREATE INDEX idx_cards_set_code ON cards (set_code)",
            "CREATE INDEX idx_cards_oracle_id ON cards (oracle_id)",
            "CREATE INDEX idx_cards_collector_number ON cards (collector_number)",
            "CREATE INDEX idx_cards_released_at ON cards (released_at)",
            "CREATE INDEX idx_cards_rarity ON cards (rarity)",
            "CREATE INDEX idx_cards_artist ON cards (artist)",
            "CREATE INDEX idx_cards_cmc ON cards (cmc)",
            "CREATE INDEX idx_cards_name_search ON cards (name COLLATE NOCASE)"
        ]
    });

    db.save(cardsCollection);

    // Create Collections collection
    const collectionsCollection = new Collection({
        "name": "collections",
        "type": "base",
        "listRule": "@request.auth.id = user.id",
        "viewRule": "@request.auth.id = user.id",
        "createRule": "@request.auth.id = user.id",
        "updateRule": "@request.auth.id = user.id",
        "deleteRule": "@request.auth.id = user.id",
        "fields": [
            {
                "name": "user",
                "type": "relation",
                "required": true,
                "maxSelect": 1,
                "collectionId": "_pb_users_auth_"
            },
            {
                "name": "card",
                "type": "relation",
                "required": true,
                "maxSelect": 1,
                "collectionId": db.findCollectionByNameOrId("cards").id
            },
            {
                "name": "quantity",
                "type": "number",
                "required": true,
                "min": 1
            },
            {
                "name": "condition",
                "type": "select",
                "required": false,
                "values": [
                    "Mint",
                    "Near Mint",
                    "Lightly Played",
                    "Moderately Played",
                    "Heavily Played",
                    "Damaged"
                ]
            },
            {
                "name": "foil",
                "type": "bool",
                "required": true
            },
            {
                "name": "acquired_date",
                "type": "date",
                "required": false
            },
            {
                "name": "notes",
                "type": "text",
                "required": false,
                "max": 1000
            }
        ]
    });

    db.save(collectionsCollection);

    const syncStatusCollection = new Collection({
        "name": "sync_status",
        "type": "base",
        "fields": [
            {
                "name": "data_type",
                "type": "select",
                "required": true,
                "values": [
                    "cards",
                    "sets",
                    "prices",
                    "images"
                ]
            },
            {
                "name": "last_sync",
                "type": "date",
                "required": false
            },
            {
                "name": "status",
                "type": "select",
                "required": true,
                "values": [
                    "success",
                    "failed",
                    "in_progress",
                    "partial"
                ]
            },
            {
                "name": "records_processed",
                "type": "number",
                "required": false,
                "min": 0
            },
            {
                "name": "error_message",
                "type": "text",
                "required": false,
                "max": 2000
            }
        ],
        "indexes": [
            "CREATE UNIQUE INDEX idx_sync_status_data_type ON sync_status (data_type)",
            "CREATE INDEX idx_sync_status_last_sync ON sync_status (last_sync)"
        ]
    });

    db.save(syncStatusCollection);
}, (db) => {
    db.deleteCollection("collections");
    db.deleteCollection("cards");
    db.deleteCollection("sync_status");
});
